README.md
# security-scripts

A collection of security related Python and Bash shell scripts, mainly revolving
around testing hosts for security vulnerabilities. For the shell scripts no
fancy programming framework is required, all that is needed is a Bash shell.

Note that it is highly recommended to use `analyze_hosts.py` as it is the most
recent version. No new features will be added to the Bash version
`analyze_hosts.sh`.

`analyze_hosts` is also available as Docker image, including the open source
scanners `droopescan`, `nmap`, `Nikto` and `testssl.sh`. Build it yourself using
the supplied Dockerfile, or grab the image from Docker Hub

```
docker run --rm gofwd/analyze_hosts
```

You can also run the included tools that way; just override the entrypoint. As
an example, run `testssl.sh`:

```
docker run --rm --entrypoint 'testssl.sh' gofwd/analyze_hosts
```

## analyze_hosts.py

A simple wrapper script around several open source security tools to simplify
scanning of hosts for network vulnerabilities. The script lets you analyze one
or several hosts for common misconfiguration vulnerabilities and weaknesses.

The main objectives for the script is to make it as easy as possible to perform
generic security tests, without any heavy prerequisites, make the output as
informative as possible, and use open source tools. It can easily be used as
scheduled task, or be implemented in Continuous Integration environments.

The only requirements are `nmap` and `Python3`.

As the scan output can be written to a JSON file it can be used to generate
deltas (differences) between scans, or to use the output for further inspection.

### Installation

Note that you can also run `analyze_hosts` straight from a Docker image:

```
docker run --rm gofwd/analyze_hosts
```

One-time installation steps without virtualenv (all required Python libraries
are specified in the `requirements.txt` file):

```
git clone https://github.com/PeterMosmans/security-scripts && \
cd security-script && \
pip3 install -r requirements.txt
```

### Usage

```
usage: analyze_hosts.py [-h] [--version] [--dry-run] [-i INPUTFILE] [-o OUTPUT_FILE] [--compact] [--queuefile QUEUEFILE] [--resume] [--settings SETTINGS]
                        [--exit-code] [--force] [--debug] [-v] [-q] [--allports] [-n] [-p PORT] [--up] [--udp] [--framework] [--http] [--compression]
                        [--headers] [--trace] [--redirect] [--force-ssl] [--json JSON] [--ssl] [--nikto] [--sslcert] [-w] [--proxy PROXY]
                        [--timeout TIMEOUT] [--threads THREADS] [--user-agent USER_AGENT] [--password PASSWORD] [--username USERNAME] [--maxtime MAXTIME]
                        [target]

analyze_hosts - scans one or more hosts for security misconfigurations

Copyright (C) 2015-2022 Peter Mosmans [Go Forward]
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

positional arguments:
  target                [TARGET] can be a single (IP) address, an IP range, or multiple comma-separated addressess

optional arguments:
  -h, --help            show this help message and exit
  --version             Show version and exit
  --dry-run             Only show commands, don't actually do anything
  -i INPUTFILE, --inputfile INPUTFILE
                        A file containing targets, one per line
  -o OUTPUT_FILE, --output-file OUTPUT_FILE
                        output file containing all scanresults (default analyze_hosts.output)
  --compact             Only log raw logfiles and alerts to file
  --queuefile QUEUEFILE
                        the queuefile
  --resume              Resume working on the queue
  --settings SETTINGS   Name of settings file to use (default analyze_hosts.yml)
  --exit-code           When supplied, return exit code 1 when alerts are discovered
  --force               Ignore / overwrite the queuefile
  --debug               Show debug information
  -v, --verbose         Be more verbose
  -q, --quiet           Do not show scan outputs on the console
  --allports            Run a full-blown nmap scan on all ports
  -n, --no-portscan     Do NOT run a nmap portscan
  -p PORT, --port PORT  Specific port(s) to scan
  --up                  Assume host is up (do not rely on ping probe)
  --udp                 Check for open UDP ports as well
  --framework           Analyze the website and run webscans
  --http                Check for various HTTP vulnerabilities (compression, headers, trace)
  --compression         Check for webserver compression
  --headers             Check for various HTTP headers
  --trace               Check webserver for HTTP TRACE method
  --redirect            Check for insecure redirect
  --force-ssl           Enforce SSL/TLS check on all open ports
  --json JSON           Save output in JSON file
  --ssl                 Check for various SSL/TLS vulnerabilities
  --nikto               Run a nikto scan
  --sslcert             Download SSL certificate
  -w, --whois           Perform a whois lookup
  --proxy PROXY         Use proxy server (host:port)
  --timeout TIMEOUT     Timeout for requests in seconds (default 10)
  --threads THREADS     Maximum number of threads (default 5)
  --user-agent USER_AGENT
                        Custom User-Agent to use (default analyze_hosts)
  --password PASSWORD   Password for HTTP basic host authentication
  --username USERNAME   Username for HTTP basic host authentication
  --maxtime MAXTIME     Timeout for scans in seconds (default 600)
```

The script `analyze_hosts` automatically execute other scans (based on their
fingerprint or open ports):

```
droopescan
nikto
testssl.sh
WPscan
```

You can use the following environment variables (all uppercase) to specify the
tools if they cannot be found in the standard paths:

CURL, DROOPESCAN, NIKTO, OPENSSL, TESTSSL, WPSCAN

### Suppressing false positives

A settings file can be used (`--settings`) to configure or tweak scan parameters
per host / port combination. This allows you to suppress false positives in scan
results. Currently the Nikto `Plugins`, `Tuning` and `output` parameters are
supported, as well as a list of allowed / expected open ports, and testssl
parameters:

Example settings file:

```
targets:
  127.0.0.1:
    allowed_ports: [22, 80, 443]
    ports:
      - port: 80
        nikto_plugins: "@@ALL"
        nikto_tuning: "x1"
        nikto_output: "report.html"
      - port: 443
        testssl_untrusted: true
        testssl:
          - "--ccs-injection"
          - "--ticketbleed"
          - "--robot"

```

This will supply the `-Plugins '@@ALL' -Tuning 'x1' -output 'report.html'
parameters to Nikto, when port 80 is scanned.

Furthermore, it will not generate an alert when an open port other than port 22,
80 or 443 is found. By default, an alert will be generated if an open port other
than 80 or 443 is found.

There will no alert be generated if the SSL/TLS endpoint on port 443 contains an
untrusted (self-signed) certificate. And instead of all default tests, only
three SSL/TLS tests will be performed.

### JSON format

```
{
  "arguments": {
    "target": "1.2.3.1/30",
    "version": false,
    "dry_run": false,
    "inputfile": "0frnfb4e",
    "output_file": "output.txt,
    "compact": true,
    "queuefile": "analyze_hosts.queue",
    "resume": false,
    "force": false,
    "debug": false,
    "verbose": false,
    "quiet": false,
    "allports": false,
    "no_portscan": false,
    "port": null,
    "up": false,
    "udp": false,
    "framework": false,
    "http": true,
    "json": "results.json",
    "ssl": true,
    "nikto": true,
    "sslcert": false,
    "trace": false,
    "whois": false,
    "proxy": null,
    "timeout": true,
    "threads": 5,
    "user_agent": "analyze_hosts",
    "password": null,
    "username": null,
    "maxtime": 1200,
    "testssl.sh": true,
    "curl": false,
    "wpscan": true,
    "droopescan": true,
    "nmap": true,
    "nmap_arguments": "-sV --open -sS --script=banner,dns-nsid,dns-recursion,http-cisco-anyconnect,http-php-version,http-title,http-trace,ntp-info,ntp-monlist,nbstat,rdp-enum-encryption,rpcinfo,sip-methods,smb-os-discovery,smb-security-mode,smtp-open-relay,ssh2-enum-algos,vnc-info,xmlrpc-methods,xmpp-info"
  },
  "date_start": "2020-05-26 31:33:06"
  "results": {
    "1.2.3.1": {
      "ports": [
        53
      ]
    },
    "1.2.3.2": {
      "ports": []
    },
    "1.2.3.3": {
      "ports": [
        80,
        443
      ],
      "alerts": [
        ":443  LUCKY13 (CVE-2013-0169), experimental     potentially VULNERABLE, uses cipher block chaining (CBC) ciphers with TLS. Check patches"
      ]
    },
    "1.2.3.4": {
      "ports": [
        80,
        443
      ],
      "alerts": [
        ":443 + OSVDB-3092: /download/: This might be interesting...",
        ":443 + OSVDB-3092: /status/: This might be interesting...",
        ":443 + OSVDB-4231: /DHrPp.xml: Coccoon from Apache-XML project reveals file system path in error messages.",
        ":443 + OSVDB-3092: /upgrade.php: upgrade.php was found."
      ]
    }
  },
  "date_finish": "2020-05-26 31:33:07"
}
```

## display_results.py

A little helper script that formats the scan results nicely, so that scan
results can easily be reviewed.

```
usage: display_results.py [-h] [--info] [--version] [inputfile]

display_results version 0.0.1 - displays scan results nicely

positional arguments:
  inputfile   A JSON file containing scan results

optional arguments:
  -h, --help  show this help message and exit
  --info      Show also informational items
  --version   Show version and exit

```

## analyze-hosts.sh

A simple wrapper script around several open source security tools to simplify
scanning of hosts for network vulnerabilities. The script lets you analyze one
or several hosts for common misconfiguration vulnerabilities and weaknesses. The
main objectives for the script is to make it as easy as possible to perform
generic security tests, without any heavy prerequisites, make the output as
informative as possible, and use open source tools....

- [cipherscan](https://github.com/jvehent/cipherscan)
- curl
- nmap
- [openssl-1.0.2-chacha](https://github.com/PeterMosmans/openssl/tree/1.0.2-chacha/)
- [whatweb](https://github.com/urbanadventurer/WhatWeb)

* whois

### Examples

#### SSL certificates

```
./analyze_hosts.sh --sslcert www.google.com
```

Shows details of a certificate, like the issuer and subject. It warns when
certificate is expired or when the certificate is a certificate authority.

Example output:

```
trying to retrieve SSL x.509 certificate on www.google.com:443... received
issuer=
    countryName               = US
    organizationName          = Google Inc
    commonName                = Google Internet Authority G2
subject=
    countryName               = US
    stateOrProvinceName       = California
    localityName              = Mountain View
    organizationName          = Google Inc
    commonName                = www.google.com
OK: certificate is valid between 16-07-2014 and 14-10-2014
```

#### SSL/TLS ciphers

```
./analyze_hosts.sh --ssl --sslports 443 -v www.microsoft.com
```

Checks which ciphers are allowed. It warns when insecure ciphers are being used.
By default the ports 443, 465, 993, 995 and 3389 and are checked. You can
specify the ports by using --sslports The -v flag outputs all results, regardles
of the message type.

Example output:

```
prio  ciphersuite   protocols    pfs_keysize
1     RC4-MD5       SSLv3,TLSv1
2     RC4-SHA       SSLv3,TLSv1
3     DES-CBC3-SHA  SSLv3,TLSv1
4     AES256-SHA    TLSv1
5     AES128-SHA    TLSv1

Certificate: UNTRUSTED, 2048 bit, sha1WithRSAEncryption signature
trying to retrieve SSL x.509 certificate on www.microsoft.com:443... received
issuer=
    domainComponent           = com
    domainComponent           = microsoft
    domainComponent           = corp
    domainComponent           = redmond
    commonName                = MSIT Machine Auth CA 2
subject=
    countryName               = US
    stateOrProvinceName       = WA
    localityName              = Redmond
    organizationName          = Microsoft Corporation
    organizationalUnitName    = MSCOM
    commonName                = www.microsoft.com
OK: certificate is valid between 12-01-2013 and 12-01-2015

performing nmap sslscan on www.microsoft.com ports 443...
Nmap scan report for www.microsoft.com (134.170.184.133)
Host is up (0.15s latency).
PORT    STATE SERVICE
443/tcp open  https
| ssl-enum-ciphers:
|   SSLv3:
|     ciphers:
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA - strong
|       TLS_RSA_WITH_RC4_128_MD5 - strong
|       TLS_RSA_WITH_RC4_128_SHA - strong
|     compressors:
|       NULL
|   TLSv1.0:
|     ciphers:
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA - strong
|       TLS_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_RSA_WITH_RC4_128_MD5 - strong
|       TLS_RSA_WITH_RC4_128_SHA - strong
|     compressors:
|       NULL
|_  least strength: strong
```

### usage

```
./analyze_hosts.sh [OPTION]... [HOST]

Scanning options:
 -a, --all               perform all basic scans
     --max               perform all advanced scans (more thorough)
 -b, --basic             perform basic scans (fingerprint, ssl, trace)
                         results of HOST matches regexp FILTER
     --dns               test for recursive query and version string
 -f                      perform web fingerprinting (all webports)
     --fingerprint       perform all web fingerprinting methods
 -h, --header            show webserver headers (all webports)
 -n, --nikto             nikto webscan (all webports)
 -p                      nmap portscan (top 1000 TCP ports)
     --ports             nmap portscan (all ports, TCP and UDP)
     --redirect          test for open secure redirect
 -s                      check SSL configuration
     --ssl               perform all SSL configuration checks
     --sslcert           show details of SSL certificate
     --timeout=SECONDS   change timeout for tools (default 60)
     --ssh               perform SSH configuration checks
 -t                      check webserver for HTTP TRACE method
     --trace             perform all HTTP TRACE method checks
 -w, --whois             perform WHOIS lookup for (hostname and) IP address
 -W                      confirm WHOIS results before continuing scan
     --filter=FILTER     only proceed with scan of HOST if WHOIS
     --wordlist=filename scan webserver for existence of files in filename

Port selection (comma separated list):
     --webports=PORTS    use PORTS for web scans (default 80,443,8080)
     --sslports=PORTS    use PORTS for ssl scans (default 443,465,993,995,3389)

Logging and input file:
 -d, --directory=DIR     location of temporary files (default /tmp)
 -i, --inputfile=FILE    use a file containing hostnames
 -l, --log               log each scan in a separate logfile
     --nocolor           don't use fancy colors in screen output
 -o, --output=FILE       concatenate all OK and WARNING messages into FILE
 -q, --quiet             quiet
 -v, --verbose           show server responses

Default programs:
     --cipherscan=FILE   location of cipherscan (default cipherscan)
     --openssl=FILE      location of openssl (default openssl)

 -u                      update this script (if it's a cloned repository)
     --update            force update (overwrite all local modifications)
     --version           print version information and exit

                         BLUE: INFO, status messages
                         GREEN: OK, secure settings
                         RED: WARNING, possible vulnerabilities

 [HOST] can be a single (IP) address, an IP range, eg. 127.0.0.1-255
 or multiple comma-separated addressess
```

### history

- since 0.88: preliminary support for starttls xmpp

## test_ssl_handhake.sh

A script to test TLS/SSL handshakes with. Several bugtests are included:

- 128 cipherlimit when using tls1_2 protocol
- aRSA cipher order
- version intolerant server

\$ ./test_ssl_handshake.sh

```
      (c) 2014 Peter Mosmans [Go Forward]
      Licensed under the GPL 3.0

tests SSL/TLS handshakes (for known bugs)

usage: ./test_ssl_handshake.sh target[:port] [start]

     [start]            number of ciphers to start with (default 125)
     --ciphers=FILE     a file containing a list which ciphers to use
     --cipherstring=CIPHERSTRING
                        cipherstring (default )
     -f | --force       continue even though the error has been detected
     --iterate          iterate through all the ciphers instead of adding
     --openssl=FILE     location of openssl (default )
     -v | --verbose     be more verbose, please

 tests:
     --128              test for 128 cipherlimit
     --intolerant       test for version intolerant server
     --rsa              test for RSA order sensitivity

     by default, all tests will be performed
```


CHANGELOG.md
# Changelog

## [1.15.0](https://git.go-forward.net/tooling/security-scripts/-/compare/1.14.0...1.15.0) (2023-01-12)

### Features

- add fours (Snap Shot Security Scanner)
  ([5603e4d](https://git.go-forward.net/tooling/security-scripts/-/commit/5603e4de2d9f2265fb7eed2f7731bb02591a9229))
- add more alert strings for Nikto and nmap scripts
  ([ae0008c](https://git.go-forward.net/tooling/security-scripts/-/commit/ae0008cdd985b747feb4cc8971908647fdc347e6))

### Bug Fixes

- don't log anything for 'empty' port numbers
  ([d077f07](https://git.go-forward.net/tooling/security-scripts/-/commit/d077f07ecf3b9a36501fed91c872872895086912))

## [1.14.0](https://git.go-forward.net/tooling/security-scripts/-/compare/1.13.0...1.14.0) (2022-10-27)

### Features

- bump version number for consistency
  ([a356a21](https://git.go-forward.net/tooling/security-scripts/-/commit/a356a21034f2974316a0817751ad670ecb71c7bd))
- change default working directory to /workdir
  ([f430cc0](https://git.go-forward.net/tooling/security-scripts/-/commit/f430cc022d3fa8bb3f1bb0b06c14d1e733c5aaca))
- hardcode name of project/Docker image
  ([a999117](https://git.go-forward.net/tooling/security-scripts/-/commit/a9991170af7e5db3225efcd275b7326e0cf6b881))
- simplify testing stage
  ([7b2662d](https://git.go-forward.net/tooling/security-scripts/-/commit/7b2662d6ccd356d2a5282c97e50075d836c8ba60))

### Bug Fixes

- suppress dockle false positive
  ([2f0957c](https://git.go-forward.net/tooling/security-scripts/-/commit/2f0957c15344c992871e188cd148bd8662a81cfe))

## [1.13.0](https://git.go-forward.net/tooling/security-scripts/-/compare/1.12.0...1.13.0) (2022-07-21)

### Features

- add basic pipeline
  ([fff2bc0](https://git.go-forward.net/tooling/security-scripts/-/commit/fff2bc0f7023926fb8b97574dd2786ae33bc4045))
- add several automated security tests
  ([0469f4f](https://git.go-forward.net/tooling/security-scripts/-/commit/0469f4fd0b71124911e74e12ebcb772b4ce70ee3))
- add workaround for wappalyzer error message
  ([b815d88](https://git.go-forward.net/tooling/security-scripts/-/commit/b815d88b4e32dcdd371c16fa3f28bde1b941537c))

### Bug Fixes

- clean up Dockerfile linting issues
  ([05c6a53](https://git.go-forward.net/tooling/security-scripts/-/commit/05c6a53f5f4f3ffb1f3c517a2d807e4a24ad0adf))
- handle connection errors gracefully
  ([8fb96fe](https://git.go-forward.net/tooling/security-scripts/-/commit/8fb96fe7fdbd76c47d453547f2f9afaff5c60359))

## [1.12.0](https://github.com/PeterMosmans/security-scripts/compare/1.11.0...1.12.0) (2022-01-11)

### Bug Fixes

- remove unused inline pylint directives
  ([0e785bf](https://github.com/PeterMosmans/security-scripts/commit/0e785bff05e14cccce56e05e9e2fcda5b828e72e))

## [1.11.0](https://github.com/PeterMosmans/security-scripts/compare/1.10.0...1.11.0) (2021-11-30)

### Features

- enforce timeout for nikto
  ([d751eaa](https://github.com/PeterMosmans/security-scripts/commit/d751eaa1a215c0487c2ccdf5726f246a5dd39438))
- ensure that Python output is unbuffered
  ([83d7fdd](https://github.com/PeterMosmans/security-scripts/commit/83d7fdd75a8228a9f17a1e32a116af36fd431b8b))
- rewrite deprecated function names
  ([e4bfbec](https://github.com/PeterMosmans/security-scripts/commit/e4bfbec7b973684c9d054b0c26dc36559739e8d8))

## [1.10.0](https://github.com/PeterMosmans/security-scripts/compare/1.9.0...1.10.0) (2021-05-27)

### Features

- support file-based argugments
  ([e33a6eb](https://github.com/PeterMosmans/security-scripts/commit/e33a6eb9bd37f2384ae2891b85355e4c966fbd5f))

## [1.9.0](https://github.com/PeterMosmans/security-scripts/compare/v1.8.0...v1.9.0) (2021-04-15)

### Features

- add multiple options to separate HTTP checks
  ([c038d28](https://github.com/PeterMosmans/security-scripts/commit/c038d28ab1d8413957e2b543d3da4bf8b2d696e8))
- add support for testssl parameters
  ([df202c0](https://github.com/PeterMosmans/security-scripts/commit/df202c00670bd7429763ad90ddfc590c7b0ed80c))
- switch protocol when trying to analyze framework
  ([ed25f28](https://github.com/PeterMosmans/security-scripts/commit/ed25f286ee87d8a3cd1737d2379c958332064449))

### Bug Fixes

- change settings YAML syntax for tuning parameters
  ([beaea06](https://github.com/PeterMosmans/security-scripts/commit/beaea06fa1eed64fdecc933eead5811354fe8a31))
- ensure that testssl untrusted parameter is properly used
  ([02dc66d](https://github.com/PeterMosmans/security-scripts/commit/02dc66d47366d7740909f1a04008b7a703789751))

## [1.8.0](https://github.com/PeterMosmans/security-scripts/compare/v1.7.0...v1.8.0) (2020-09-18)

### Features

- add option to enforce SSL/TLS check
  ([b641f7a](https://github.com/PeterMosmans/security-scripts/commit/b641f7a78ab2d4579d016d8821ab3c6c3c6836c7))

### Bug Fixes

- ensure that a host contains a port key
  ([6083104](https://github.com/PeterMosmans/security-scripts/commit/6083104f28598f7ea71ccb557557c0e7bbebe0c9))

## [1.7.0](https://github.com/PeterMosmans/security-scripts/compare/v1.6.0...v1.7.0) (2020-08-24)

### Features

- generate alert when unexpected open port is found
  ([9cbc4a4](https://github.com/PeterMosmans/security-scripts/commit/9cbc4a458ece2a213d05cfd1ddd446c56ed1fb1c))
- **display_hosts:** add several new options
  ([6c7077e](https://github.com/PeterMosmans/security-scripts/commit/6c7077e9be6c0b7e6e9e35ef0c781b277d65bfb3))
- add display_results helper script
  ([6016cab](https://github.com/PeterMosmans/security-scripts/commit/6016cab79f6a6020929b6479dca10e6ad3e83206))
- improve displaying results, show unexpected open ports
  ([5f8a8af](https://github.com/PeterMosmans/security-scripts/commit/5f8a8aff01a86d66fb340f630f603e07a53d30ee))

### Bug Fixes

- ensure results are shown without specifying host
  ([0efe1c5](https://github.com/PeterMosmans/security-scripts/commit/0efe1c573c9a569b6ece19be9bdf2113c407ddee))

## [1.6.0](https://github.com/PeterMosmans/security-scripts/compare/v1.5.0...v1.6.0) (2020-08-06)

### Features


fours.sh
#!/usr/bin/env bash

# Fours - Snap Shot Security Scanner
# Perform a security scan given a set of parameters, and report in HTML

# Copyright (C) 2023, Peter Mosmans
# SPDX-License-Identifier: GPL-3.0-only

IMAGE=gofwd/analyze_hosts
TIMEZONE=Europe/Amsterdam
TARGETS=targets.txt
SCANLOG=scan-results.txt
SCANRESULTS=scan-results.json
HTMLRESULTS=scan-results.html

if ! which docker &>/dev/null; then
    echo "Docker not found or not installed - exiting prematurely"
    exit 1
fi

if [[ ! -f $TARGETS ]]; then
    echo "Target file $TARGETS not found - exiting prematurely"
    exit 1
fi

echo "Fours v0.1 - Snap Shot Security Scanner - (c) 2023 Peter Mosmans"
if [[ -f analyze_hosts.queue ]]; then
    echo -n "It looks like a scan is still active: "
    state=$(docker ps --format '{{.Image}} Created {{.RunningFor}} - {{.Status}}' | grep $IMAGE)
    echo "$state"
    if [[ -z "$state" ]]; then
        echo "Hmm... it seems $IMAGE is not running (anymore)"
        echo "You might want to manually remove analyze_hosts.queue and restart the script"
    fi
    echo "Exiting prematurely"
    exit 0
fi

echo "Starting scan of $(wc -l $TARGETS) target(s)"
if [[ ! -f $SCANLOG ]]; then
    echo "Appending scan log to $SCANLOG"
fi

docker run --rm --volume "$PWD":/workdir -e TZ=$TIMEZONE $IMAGE \
--compact \
--http \
--nikto \
--quiet \
--settings analyze_hosts.yml \
--ssl \
--threads 10 \
-i $TARGETS \
-o $SCANLOG \
--json $SCANRESULTS

if [[ ! -f $SCANRESULTS ]]; then
    echo "$SCANRESULTS not created - exiting prematurely"
    exit 1
fi

echo "Converting scan results into HTML file $HTMLRESULTS"
docker run --rm --volume "$PWD":/workdir --entrypoint '' gofwd/analyze_hosts \
results_to_html.py $SCANRESULTS > $HTMLRESULTS


results_to_html.py
#!/usr/bin/env python3

"""results_to_html - Convert JSON results to a simple HTML file

Copyright (C) 2023 Peter Mosmans [Go Forward]
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
"""

import ipaddress
import json
import sys
from jinja2 import Environment, FileSystemLoader
import typer

app = typer.Typer()


def sorted_hosts(unsorted):
    """Return a list of hosts sorted on IP address (when possible)."""
    try:
        results = map(str, sorted([ipaddress.ip_address(x) for x in unsorted]))
    except:
        results = sorted(unsorted)
    return results


@app.command()
def convert(filename: str, templates="/usr/share/templates"):
    """Convert JSON results to HTML."""
    try:
        with open(filename, encoding="utf-8") as handle:
            results = json.load(handle)
        hosts = sorted_hosts(results["results"])
        environment = Environment(loader=FileSystemLoader(templates))
        template = environment.get_template("results.html")
        print(template.render(results=results, hosts=hosts))
    except Exception as e:
        print(f"Something went wrong: {e}", file=sys.stderr)
        sys.exit(-1)


if __name__ == "__main__":
    app()


display_results.py
#!/usr/bin/env python3

import argparse
import ipaddress
import json
import sys
import textwrap

try:
    from colorama import Fore, Style
except ImportError as exception:
    print(
        f"[-] Please install required modules, e.g. pip3 install -r requirements.txt: {exception}",
        file=sys.stderr,
    )

NAME = "display_results"
VERSION = "0.3.0"
ALLOWED_PORTS = [80, 443]


def parse_arguments(banner):
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description=textwrap.dedent(
            banner
            + """\
 - displays scan results nicely

Copyright (C) 2020  Peter Mosmans [Go Forward]
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version."""
        ),
    )
    parser.add_argument(
        "inputfile",
        nargs="?",
        action="store",
        type=str,
        help="A JSON file containing scan results",
    )
    parser.add_argument(
        "--empty", action="store_true", help="Show hosts without any results",
    )
    parser.add_argument(
        "--host", action="store", help="Filter on specific host",
    )
    parser.add_argument(
        "--info",
        action="store_true",
        help="Show also informational items",
        default=False,
    )
    parser.add_argument(
        "--ports",
        action="store_true",
        help=f"Show ports other than {ALLOWED_PORTS} that are open",
    )
    parser.add_argument("--version", action="store_true", help="Show version and exit")
    args = parser.parse_args()
    if args.version:
        print(banner)
        sys.exit(0)
    if not args.inputfile:
        parser.error("Specify a JSON input file")
    return vars(parser.parse_args())


def sorted_hosts(unsorted):
    """Return a list of hosts sorted on IP address."""
    return map(str, sorted([ipaddress.ip_address(x) for x in unsorted]))


def format_alert(alert):
    """Return a string formatted as an alert."""
    return f"{Fore.GREEN}{alert}{Fore.RESET}"


def display_json(filename, **options):
    """Display filename sorted on IP address."""
    try:
        with open(filename, mode="r", encoding="utf-8") as json_file:
            results = json.load(json_file)
        print(
            f"Scan of {results['arguments']['target']} was started at {Style.BRIGHT}{results['date_start']}{Style.NORMAL}"
        )
        hosts = results["results"]
        targets = sorted_hosts(hosts)
        for target in targets:
            if options["host"] and target != options["host"]:
                continue
            result = {}
            for port in hosts[target]["ports"]:
                result[str(port)] = []
            if "info" in hosts[target] and options["info"]:
                for port in hosts[target]["info"]:
                    for item in hosts[target]["info"][port]:
                        result[port].append(item)
            if "alerts" in hosts[target]:
                for port in hosts[target]["alerts"]:
                    for item in hosts[target]["alerts"][port]:
                        result[port].append(format_alert(item))
            if options["empty"] or len(result) > 1:
                print(f"{Style.BRIGHT}{target}{Style.NORMAL}")
                for port, items in result.items():
                    if (
                        not len(items)
                        and options["ports"]
                        and (int(port) not in ALLOWED_PORTS)
                    ):
                        print(f" {Fore.MAGENTA}{port:>5}{Fore.RESET} OPEN PORT")
                    for item in items:
                        print(f" {Fore.MAGENTA}{port:>5}{Fore.RESET} {item}")
                        port = ""  # Only show port number for the first row
                print("")
    except FileNotFoundError as exception:
        print(f"File {filename} could not be found: Exception {exception}")
    except KeyError as exception:
        print(f"File {filename} not in expected format: Exception {exception}")


def main():
    """Main program loop."""
    banner = "{0} version {1}".format(NAME, VERSION)
    options = parse_arguments(banner)
    display_json(options["inputfile"], **options)
    sys.exit(0)


if __name__ == "__main__":
    main()
